task/sensor-types/src/lib.rs

```diff
@@ -0,0 +1,158 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Client types for the Sensor API
+#![no_std]
+
+use derive_idol_err::IdolError;
+use drv_i2c_types::ResponseCode;
+use hubpack::SerializedSize;
+use num_derive::FromPrimitive;
+use num_traits::FromPrimitive;
+use serde::{Deserialize, Serialize};
+
+#[derive(
+    zerocopy::AsBytes,
+    Copy,
+    Clone,
+    Debug,
+    Eq,
+    PartialEq,
+    Serialize,
+    Deserialize,
+    SerializedSize,
+)]
+#[repr(C)]
+pub struct SensorId(pub u32);
+
+impl From<u32> for SensorId {
+    fn from(id: u32) -> Self {
+        SensorId(id)
+    }
+}
+
+impl From<SensorId> for u32 {
+    fn from(id: SensorId) -> Self {
+        id.0
+    }
+}
+
+#[derive(Copy, Clone, Debug, SerializedSize, Serialize, Deserialize)]
+pub struct Reading {
+    pub timestamp: u64,
+    pub value: f32,
+}
+
+impl Reading {
+    pub fn new(value: f32, timestamp: u64) -> Self {
+        Self { timestamp, value }
+    }
+}
+
+//
+// Note that [`counter_encoding`] relies on [`NoData`] being numbered from 0 and
+// being numbered sequentially.
+//
+#[derive(
+    zerocopy::AsBytes,
+    Copy,
+    Clone,
+    Debug,
+    FromPrimitive,
+    Eq,
+    PartialEq,
+    Serialize,
+    Deserialize,
+    SerializedSize,
+)]
+#[repr(u8)]
+pub enum NoData {
+    DeviceOff = 0,
+    DeviceError = 1,
+    DeviceNotPresent = 2,
+    DeviceUnavailable = 3,
+    DeviceTimeout = 4,
+}
+
+impl NoData {
+    ///
+    /// Routine to determine the number of bits and size of shift
+    /// to pack a counter for each [`NoData`] variant into type `T`.
+    ///
+    pub fn counter_encoding<T>(self) -> (usize, usize) {
+        //
+        // We need to encode the number of variants in [`NoData`] here.  There
+        // is a very convenient core::mem::variant_count() that does exactly
+        // this, but it's currently unstable -- so instead we have an
+        // exhaustive match to assure that the enum can't be updated without
+        // modifying this code.
+        //
+        let nbits = (core::mem::size_of::<T>() * 8)
+            / match self {
```

Comment:
Nit - maybe this is a place for `strum` (specifically [`EnumCount`](https://docs.rs/strum/latest/strum/trait.EnumCount.html))?

Reply:
This was copy-pasta extracted from `sensor-api`, so I don't think I'll change it in this PR.